firmware/examples/FactoryTest/source/main.cpp

```diff
@@ -0,0 +1,323 @@
+  bool result = false;
+  cs.SetAsOutput();
+  cs.SetHigh();
+
+  ssp2.SetPeripheralMode(Ssp::MasterSlaveMode::kMaster, Ssp::FrameMode::kSpi,
+                         Ssp::DataSize::kEight);
+  ssp2.SetClock(false, false, 1, 48);
+  ssp2.Initialize();
+  Delay(1000);
+
+  uint8_t array[5];
+  cs.SetLow();
+  Delay(1);
+  // Read Manufacturer ID
+  ssp2.Transfer(0x9F);
+  array[0] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  array[1] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  array[2] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  array[3] = static_cast<uint8_t>(ssp2.Transfer(0x00));
+  LOG_INFO("Returned 0x%02X 0x%02X 0x%02X 0x%02X", array[0], array[1], array[2],
+           array[3]);
+  cs.SetHigh();
+  Delay(1000);
+  LOG_INFO("End of External Flash Test.");
+  if ((array[0] == 0x1F) && (array[1] == 0x40) && (array[2] == 0x00) &&
+      (array[3] == 0x00))
+  {
+    result = true;
+  }
+  return result;
+}
+
+bool I2cTest()
+{
+  // I2C device scan.
+  i2c.Initialize();
+  // Creates a status variable for I2C transactions
+  Status status;
+  int devices_found               = 0;
+  constexpr int total_i2c_devices = 3;
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Beginning I2c scan...\n\n");
+  for (uint8_t address = kFirstI2cAddress; address < kLastI2cAddress; address++)
+  {
+    status = i2c.Write(address, nullptr, 0);
+    if (status == Status::kSuccess)
+    {
+      LOG_INFO("    Found device at address: 0x%02X", address);
+      if (address == kAccelerometerAddress || address == kTemperatureAddress ||
+          address == kGestureAddress)
+      {
+        devices_found++;
+      }
+    }
+  }
+  bool i2c_pass           = (devices_found == total_i2c_devices);
+  bool gesture_pass       = GestureSensorTest();
+  bool accelerometer_pass = AccelerometerTest();
+  bool temp_pass          = TemperatureSensorTest();
+  bool final_result =
+      gesture_pass && accelerometer_pass && temp_pass && i2c_pass;
+  return final_result;
+}
+
+bool GpioFunctionalityTest()
+{
+  // P0_2 and P0_3 do not need to be tested.
+  // P0_29 and P0_30 do not need to be tested (lack of pullup/pulldown)
+  constexpr uint8_t num_ports = 5;
+  constexpr uint8_t num_pins  = 32;
+
+  const bool kLqfP80[num_ports][num_pins] = {
+    { 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1,
+      1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0 },
+    { 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1,
+      0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1 },
+    { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0 }
+  };
+
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Begin Gpio test...\n\n");
+
+  uint8_t gpio_total                  = 0;
```

Comment:
```suggestion
  constexpr uint8_t kGpioTotal                  = GetTotalOfGpiosToTest(kLQFP80);
```
This should be constexpr. You can make kLQFP80 a constexpr array, pass that to a constexpr function, that will return the total by checking all of the 1s in the bitmap.